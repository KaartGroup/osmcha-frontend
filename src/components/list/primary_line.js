// @flow
import React from 'react';

export function PrimaryLine({ comment }: { comment: String }) {
  return (
    <span className="flex-parent flex-parent--column">
      <p className="flex-child truncate-3-lines my6">
        {comment}
      </p>
    </span>
  );
}
