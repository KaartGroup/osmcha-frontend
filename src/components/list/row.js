// @flow
import React from 'react';
import { Link } from 'react-router-dom';
import { Map } from 'immutable';

import { SecondaryLine } from './secondary_line';
import { PrimaryLine } from './primary_line';
import { Title } from './title';

export class Row extends React.PureComponent {
  props: {
    properties: Map<string, *>,
    active: ?boolean,
    changesetId: number,
    inputRef: ?(any) => any
  };
  shouldComponentUpdate(nextProps: Object) {
    return (
      nextProps.properties !== this.props.properties ||
      this.props.active ||
      nextProps.active
    );
  }
  wasOpen = false;
  render() {
    console.log('rendering', this.props.changesetId);
    const { properties, changesetId, active, inputRef, ...other } = this.props;
    if (!this.wasOpen) {
      // way to show read/unread state without
      // performance compromise. The moment component
      // gets active we set wasOpen to true and never
      // toggle it back to any other state.
      this.wasOpen = this.props.active;
    }
    return (
      <Link to={`/changesets/${changesetId}`}>
        <div
          className={`${active ? 'bg-green-faint bg-green-faint-on-hover' : ' bg-gray-faint-on-hover '} transition`}
          ref={inputRef}
        >
          <div
            {...other}
            className={
              'ml12 cursor-pointer flex-parent flex-parent--row justify--space-between border-b py6 border-b--1 border--gray-light'
            }
          >
            <div className="flex-parent flex-parent--row">
              <div className="txt-mono">{this.wasOpen ? '\u00a0' : 'â€¢'}</div>
              <div className="flex-parent flex-parent--column">
                <div>
                  <Title properties={properties} wasOpen={this.wasOpen} />
                </div>
                <div>
                  <PrimaryLine comment={properties.get('comment')} />
                </div>
                <div>
                  <SecondaryLine
                    changesetId={changesetId}
                    date={properties.get('date')}
                  />
                </div>
              </div>
            </div>
            <div />
          </div>
        </div>
      </Link>
    );
  }
}
