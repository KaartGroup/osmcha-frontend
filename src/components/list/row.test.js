import React from 'react';
import renderer from 'react-test-renderer';
import { fromJS } from 'immutable';
import { StaticRouter } from 'react-router';
import { PrimaryLine } from './primary_line';
import { Row } from './row';
import MockDate from 'mockdate';

const changeset = fromJS({
  id: 49328744,
  type: 'Feature',
  geometry: {
    type: 'Polygon',
    coordinates: [
      [
        [-3.8867418, 15.0599119],
        [-3.874586, 15.0599119],
        [-3.874586, 15.0771946],
        [-3.8867418, 15.0771946],
        [-3.8867418, 15.0599119]
      ]
    ]
  },
  properties: {
    check_user: 'nammala',
    reasons: [
      { id: 4, name: 'mass deletion' },
      { id: 72, name: 'Randomized flag' }
    ],
    tags: [{ id: 1, name: 'intentional' }, { id: 2, name: 'unintentional' }],
    features: [
      {
        osm_id: 498804541,
        url: 'way-498804541',
        name: null,
        reasons: [{ id: 72, name: 'Randomized flag' }]
      }
    ],
    user: 'DaryR',
    uid: '4569402',
    editor: 'JOSM/1.5 (10966 en)',
    comment: '#hotosm-project-2999 #MissingMaps #EndMalaria #Mali',
    source:
      'tms[19]:http://{switch:a,b,c}.tiles.mapbox.com/v4/digitalglobe.316c9a2e/{zoom}/{x}/{y}.png?access_token=pk.eyJ1IjoiZGlnaXRhbGdsb2JlIiwiYSI6ImNqMmFxdGp5aTAwOWIzM3M1NDZ2eGU1a2QifQ.JN5adNxCwK_oprEwFEtjjg',
    imagery_used: 'Not reported',
    date: '2017-06-07T08:25:38Z',
    create: 1624,
    modify: 26,
    delete: 2666,
    area: 0.00021008504466001,
    is_suspect: true,
    harmful: true,
    checked: true,
    check_date: '2017-06-08T08:51:29.983657Z'
  }
});
it('renders PrimaryLine correctly', () => {
  MockDate.set(1497172627326);

  const tree = renderer
    .create(
      <PrimaryLine
        reasons={changeset.getIn(['properties', 'reasons'])}
        tags={changeset.getIn(['properties', 'tags'])}
        comment={changeset.getIn(['properties', 'comment'])}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
  MockDate.reset();
});

it('renders active row properly', () => {
  MockDate.set(1497172627326);

  const tree1 = renderer
    .create(
      <StaticRouter>
        <Row
          properties={changeset.getIn(['properties'])}
          active
          changesetId={changeset.getIn(['id'])}
          inputRef={() => {}}
        />
      </StaticRouter>
    )
    .toJSON();
  expect(tree1).toMatchSnapshot();
  MockDate.reset();
});

it('renders inactive row properly', () => {
  MockDate.set(1497172627326);

  const tree1 = renderer
    .create(
      <StaticRouter>
        <Row
          properties={changeset.getIn(['properties'])}
          active={false}
          changesetId={changeset.getIn(['id'])}
          inputRef={() => {}}
        />
      </StaticRouter>
    )
    .toJSON();
  expect(tree1).toMatchSnapshot();
  MockDate.reset();
});
