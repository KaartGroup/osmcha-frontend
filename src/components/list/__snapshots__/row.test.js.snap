// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PrimaryLine correctly 1`] = `
<span
  className="flex-parent flex-parent--column"
>
  <p
    className="flex-child truncate-3-lines my6"
  >
    #hotosm-project-2999 #MissingMaps #EndMalaria #Mali
  </p>
  <span>
    <span>
      <span
        className="color-red-dark bg-red-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
      >
        intentional
      </span>
      , 
    </span>
    <span>
      <span
        className="color-red-dark bg-red-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
      >
        unintentional
      </span>
      
    </span>
  </span>
  <span>
    <span>
      <span
        className="color-green-dark bg-green-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
      >
        mass deletion
      </span>
      , 
    </span>
    <span>
      <span
        className="color-green-dark bg-green-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
      >
        Randomized flag
      </span>
      
    </span>
  </span>
</span>
`;

exports[`renders active row properly 1`] = `
<a
  href="/changesets/49328744"
  onClick={[Function]}
>
  <div
    className="light-blue border-l border-l--4 border-color-bad"
  >
    <div
      className="p12 cursor-pointer flex-parent flex-parent--column border-b border-b--1 border--gray-light"
    >
      <div
        className="flex-parent flex-parent--column"
      >
        <div>
          <div>
            <span
              className="flex-parent flex-parent--row justify--space-between align-items--center"
            >
              <span
                className="txt-m  mt3 mr6"
              >
                DaryR
              </span>
              <span
                className="txt-s mr3"
              >
                Â 
                4 days ago
              </span>
            </span>
          </div>
        </div>
        <div>
          <span
            className="flex-parent flex-parent--column"
          >
            <p
              className="flex-child truncate-3-lines my6"
            >
              #hotosm-project-2999 #MissingMaps #EndMalaria #Mali
            </p>
            <span>
              <span>
                <span
                  className="color-red-dark bg-red-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
                >
                  intentional
                </span>
                , 
              </span>
              <span>
                <span
                  className="color-red-dark bg-red-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
                >
                  unintentional
                </span>
                
              </span>
            </span>
            <span>
              <span>
                <span
                  className="color-green-dark bg-green-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
                >
                  mass deletion
                </span>
                , 
              </span>
              <span>
                <span
                  className="color-green-dark bg-green-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
                >
                  Randomized flag
                </span>
                
              </span>
            </span>
          </span>
        </div>
        <div>
          <span
            className="flex-parent flex-parent--row justify--space-between txt-light txt-s color-gray"
          >
            <span>
              <span>
                49328744
              </span>
              <span>
                ðŸ‘Ž 
                Â  byÂ 
                nammala
              </span>
            </span>
            <span
              className="flex-parent flex-parent--row"
            >
              <div
                className="flex-parent flex-parent--row flex-parent--wrap mr3"
              >
                <div
                  className="cmap-bg-create  color-white inline-block px6 txt-s txt-bold round unround-r"
                >
                  1624
                </div>
                <div
                  className="cmap-bg-modify-old  color-white inline-block px6  txt-s txt-bold unround"
                >
                  26
                </div>
                <div
                  className="cmap-bg-delete color-white inline-block px6 txt-s txt-bold round unround-l"
                >
                  2666
                </div>
              </div>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</a>
`;

exports[`renders inactive row properly 1`] = `
<a
  href="/changesets/49328744"
  onClick={[Function]}
>
  <div
    className=" border-l border-l--4 border-color-bad"
  >
    <div
      className="p12 cursor-pointer flex-parent flex-parent--column border-b border-b--1 border--gray-light"
    >
      <div
        className="flex-parent flex-parent--column"
      >
        <div>
          <div>
            <span
              className="flex-parent flex-parent--row justify--space-between align-items--center"
            >
              <span
                className="txt-m txt-bold mt3 mr6"
              >
                DaryR
              </span>
              <span
                className="txt-s mr3"
              >
                Â 
                4 days ago
              </span>
            </span>
          </div>
        </div>
        <div>
          <span
            className="flex-parent flex-parent--column"
          >
            <p
              className="flex-child truncate-3-lines my6"
            >
              #hotosm-project-2999 #MissingMaps #EndMalaria #Mali
            </p>
            <span>
              <span>
                <span
                  className="color-red-dark bg-red-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
                >
                  intentional
                </span>
                , 
              </span>
              <span>
                <span
                  className="color-red-dark bg-red-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
                >
                  unintentional
                </span>
                
              </span>
            </span>
            <span>
              <span>
                <span
                  className="color-green-dark bg-green-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
                >
                  mass deletion
                </span>
                , 
              </span>
              <span>
                <span
                  className="color-green-dark bg-green-faint inline-block px3 txt-mono mr3 txt-s txt-bold txt-underline"
                >
                  Randomized flag
                </span>
                
              </span>
            </span>
          </span>
        </div>
        <div>
          <span
            className="flex-parent flex-parent--row justify--space-between txt-light txt-s color-gray"
          >
            <span>
              <span>
                49328744
              </span>
              <span>
                ðŸ‘Ž 
                Â  byÂ 
                nammala
              </span>
            </span>
            <span
              className="flex-parent flex-parent--row"
            >
              <div
                className="flex-parent flex-parent--row flex-parent--wrap mr3"
              >
                <div
                  className="cmap-bg-create  color-white inline-block px6 txt-s txt-bold round unround-r"
                >
                  1624
                </div>
                <div
                  className="cmap-bg-modify-old  color-white inline-block px6  txt-s txt-bold unround"
                >
                  26
                </div>
                <div
                  className="cmap-bg-delete color-white inline-block px6 txt-s txt-bold round unround-l"
                >
                  2666
                </div>
              </div>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</a>
`;
